File: 2022/practice/graph/0310-minimum-height-trees.go
// 两种方法：
// 1）找到图中的最长路径，处于路径上中间位置的节点可以使得高度为路径的二分之一
// 2）拓扑排序
func findMinHeightTrees(n int, edges [][]int) []int {
    if n == 1 {
        return []int{0}
    }

    // 建立邻接表以及度统计
    g := make([][]int, n)
    deg := make([]int, n)
    for i := range edges {
        src, des := edges[i][0], edges[i][1]
        g[src] = append(g[src], des)
        g[des] = append(g[des], src)
        deg[src]++
        deg[des]++
    }

    // 统计度为1的节点，这种叶子结点一定是在最长路径两端
    var q []int
    for i := range deg {
        if deg[i] == 1 {
            q = append(q, i)
        }
    }

    // 进行拓扑排序
    remainNodeCnt := n 
    for remainNodeCnt > 2 {
        remainNodeCnt -= len(q)
        tmp := q
        q = nil

        for _, v := range tmp {
            for _, n := range g[v] {
                deg[n]--
                if deg[n] == 1 {
                    q = append(q, n)
                }
            }
        }
    }

    return q
}

File: 2022/practice/graph/0310-minimum-height-trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n == 1:
            return [0]
        
        g = [[] for _ in range(n)]
        for e in edges:
            src, des = e[0], e[1]
            g[src].append(des)
            g[des].append(src)
            
        parent = [0] * n
        maxL, node = 0, -1
        def dfs(x: int, p: int, l: int):
            nonlocal maxL, node
            if l > maxL:
                maxL, node = l, x
            parent[x] = p

            for y in g[x]:
                if y != p:
                    dfs(y, x, l+1)
        
        dfs(0, -1, 1)
        maxL = 0
        dfs(node, -1, 1)

        path = []
        while node != -1:
            path.append(node)
            node = parent[node]
        
        l = len(path)
        if l % 2 == 1:
            return path[l//2:l//2+1]

        return path[l//2-1:l//2+1]
